MATLAB/+Battery/InputEqnv1.m

```diff
@@ -0,0 +1,28 @@
+function Output = InputEqn(parameters,t,inputParameters)
+
+persistent input inputparameters time;
+
+if isempty(input) || ~isequal(inputparameters,inputParameters)
+    inputparameters=inputParameters;
+    power=inputparameters(1:2:end,:);
+    duration=inputparameters(2:2:end,:);
+    N=size(power,2);
+    input=zeros(round(max(sum(duration,1))/parameters.sampleTime),N);
+    time=0:parameters.sampleTime:size(input,1)*parameters.sampleTime;
+    for j=1:N
+        temp=[];
+        for i=1:size(power,1)
+            temp=[temp; repmat(power(i,j),round(duration(i,j)/parameters.sampleTime),1) ];
+        end
+        input(1:length(temp),j)=temp;
+        input(length(temp)+1:end,j)=temp(end);
+    end

```

Comment:
The way this works is that a time-based matrix is constructed in the first call to InputEqn (or when inputParameters changes). Precomputing this matrix gives the speedup, because then it is not recomputed at each call to InputEqn. This is the first difference with the original InputEqn.

The second difference is that the matrix used to lookup the load for the current time is time-based, instead of segment-based. For example consider 10 segments, covering 1 hour, with a sample time for the model of 1 second, with 100 prediction samples. The time-based matrix would be 3600x100. For a sampling time of 0.1 seconds, it is 36000x100. This is not scalable and would result in memory and performance issues if the sampling time is decreased and/or the total duration covered by the segments is increased.

In contrast, the segment-based matrix will always be of size (n_segments)x(n_samples). So I would suggest that instead of precomputing the time-based matrix, precompute the segment-based matrix (the 'loads' and 'times' variables in the original InputEqn). Then this would be a more straightforward extension of the original InputEqn where those are precomputed on the first call to InputEqn (or when inputParameters changes) and should be even more efficient than this revision.